libwebcash/src/webcash.c

```diff
@@ -5,30 +5,299 @@
+    g_wallet.hdroot.chain[MINING].maxdepth = 0;
+    g_wallet.hdroot.chain[MINING].keys = NULL;
+}
+
+static void BytesToHexString(char* hex, const uint8_t* bytes, size_t len) {
+    char hexmap[] = "0123456789abcdef";
+    for (size_t i = 0; i < len; i++) {
+        hex[2*i] = hexmap[(bytes[i] >> 4) & 0xf];
+        hex[2*i + 1] = hexmap[bytes[i] & 0xf];
+    }
+}
+
+struct webcash_wallet_secret* webcash_wallet_reserve_secret(uint64_t timestamp, bool mine, bool sweep) {
+    if (!g_is_initialized) {
+        g_wallet.callbacks.raise("webcash_wallet_reserve_secret() called before webcash_init()");
+        return NULL; // Never reached
+    }
+
+    struct webcash_wallet_hdchain* chain = NULL;
+    const uint64_t chaincode = 0;
+    uint8_t chaincode_bytes[8] = {
+        (uint8_t) (chaincode >> 54) & 0xff,
+        (uint8_t) (chaincode >> 46) & 0xff,
+        (uint8_t) (chaincode >> 38) & 0xff,
+        (uint8_t) (chaincode >> 30) & 0xff,
+        (uint8_t) (chaincode >> 22) & 0xff,
+        (uint8_t) (chaincode >> 14) & 0xff,
+        (uint8_t) (chaincode >> 6) & 0xff,
+        (uint8_t) (chaincode << 2) & 0xfc,
+    };
+    if (!mine && sweep) {
+        chaincode_bytes[7] |= 0; // RECEIVE
+        chain = &g_wallet.hdroot.chain[RECEIVE];
+    } else if (!mine && !sweep) {
+        chaincode_bytes[7] |= 1; // PAYMENT
+        chain = &g_wallet.hdroot.chain[PAYMENT];
+    } else if (mine && !sweep) {
+        chaincode_bytes[7] |= 2; // CHANGE
+        chain = &g_wallet.hdroot.chain[CHANGE];
+    } else if (mine && sweep) {
+        chaincode_bytes[7] |= 3; // MINING
+        chain = &g_wallet.hdroot.chain[MINING];
+    }
+    uint64_t depth = chain->maxdepth++;
+    uint8_t depth_bytes[8] = {
+        (uint8_t) (depth >> 56) & 0xff,
+        (uint8_t) (depth >> 48) & 0xff,
+        (uint8_t) (depth >> 40) & 0xff,
+        (uint8_t) (depth >> 32) & 0xff,
+        (uint8_t) (depth >> 24) & 0xff,
+        (uint8_t) (depth >> 16) & 0xff,
+        (uint8_t) (depth >> 8) & 0xff,
+        (uint8_t) (depth >> 0) & 0xff,
+    };
+
+    // Secret derivation uses a tagged-hash function, which is a sha256 context
+    // initialized with the hash of the tag string, repeated twice.
+    static const char *tag_str = "webcashwalletv1";
+    struct sha256_ctx ctx;
+    sha256_init(&ctx);
+    sha256_update(&ctx, tag_str, strlen(tag_str));
+    struct sha256 tag;
+    sha256_done(&tag, &ctx);
+
+    sha256_init(&ctx);
+    sha256_update(&ctx, tag.u8, sizeof(tag.u8));
+    sha256_update(&ctx, tag.u8, sizeof(tag.u8));
+    sha256_update(&ctx, chaincode_bytes, sizeof(chaincode_bytes));
+    sha256_update(&ctx, depth_bytes, sizeof(depth_bytes));
+    struct sha256 secret;
+    sha256_done(&secret, &ctx);
+
+    struct webcash_wallet_hdkey* hdkey = malloc(sizeof(struct webcash_wallet_hdkey));
+    if (hdkey == NULL) {
+        g_wallet.callbacks.raise("webcash_wallet_reserve_secret() failed to allocate memory");
+        return NULL; // Never reached
```

Comment:
My suggestion is to make a function that calls this one, which has the `__attribute__((__noreturn__))` attribute attached. In `.h` file:
```C
void webcash_raise_exception(const char* fmt, ...) __attribute__((__noreturn__));
```
In `.c` file:
```C
void webcash_raise_exception(const char* fmt, ...)
{
    static char buf[1024];
    va_list ap;
    va_start(ap, fmt);
    vsnprintf(&buf, 1024, fmt, ap);
    va_end(ap);
    g_wallet.callbacks.raise(&buf);
    exit(1);
}
```

Reply:
(This is untested code; I was leaking `buf` in the original and swapped it into a static non-pointy char array here.)